demos/kdf/argon2.c

```diff
@@ -0,0 +1,156 @@
+};
+
+/*
+ * Argon2 cost parameters
+ */
+static uint32_t memory_cost    = 32;
+static uint32_t iteration_cost =  3;
+static uint32_t parallel_cost  =  4;
+
+static const unsigned char expected_output[] = {
+   0x0d, 0x64, 0x0d, 0xf5, 0x8d, 0x78, 0x76, 0x6c,
+   0x08, 0xc0, 0x37, 0xa3, 0x4a, 0x8b, 0x53, 0xc9,
+   0xd0, 0x1e, 0xf0, 0x45, 0x2d, 0x75, 0xb6, 0x5e,
+   0xb5, 0x25, 0x20, 0xe9, 0x6b, 0x01, 0xe6, 0x59
+};
+
+int main(int argc, char **argv)
+{
+    int rv = 1;
+    EVP_KDF *kdf = NULL;
+    EVP_KDF_CTX *kctx = NULL;
+    unsigned char out[32];
+    OSSL_PARAM params[8], *p = params;
+    OSSL_LIB_CTX *library_context = NULL;
+    unsigned int threads;
+
+    library_context = OSSL_LIB_CTX_new();
+    if (library_context == NULL) {
+        fprintf(stderr, "OSSL_LIB_CTX_new() returned NULL\n");
+        goto end;
+    }
+
+    /* Fetch the key derivation function implementation */
+    kdf = EVP_KDF_fetch(library_context, "argon2id", NULL);
+    if (kdf == NULL) {
+        fprintf(stderr, "EVP_KDF_fetch() returned NULL\n");
+        goto end;
+    }
+
+    /* Create a context for the key derivation operation */
+    kctx = EVP_KDF_CTX_new(kdf);
+    if (kctx == NULL) {
+        fprintf(stderr, "EVP_KDF_CTX_new() returned NULL\n");
+        goto end;
+    }
+
+    /*
+     * Thread support can be turned off; use serialization if we cannot
+     * set requested number of threads.
+     */
+    threads = parallel_cost;
+    if (OSSL_set_max_threads(library_context, parallel_cost) != 1) {
+        uint64_t max_threads = OSSL_get_max_threads(library_context);
+
+        if (max_threads == 0)
+            threads = 1;
+        else if (max_threads < parallel_cost)
+            threads = max_threads;
+    }
+
+    /* Set password */
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_PASSWORD, password, sizeof(password));
+    /* Set salt */
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_SALT, salt, sizeof(salt));
+    /* Set optional additional data */
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_ARGON2_AD, ad, sizeof(ad));
+    /* Set optional secret */
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_SECRET, secret, sizeof(secret));
+    /* Set iteration count */
+    *p++ = OSSL_PARAM_construct_uint32(OSSL_KDF_PARAM_ITER, &iteration_cost);
+    /* Set threads performing derivation (can be decreased) */
+    *p++ = OSSL_PARAM_construct_uint(OSSL_KDF_PARAM_THREADS, &threads);
+    /* Set parallel cost */
+    *p++ = OSSL_PARAM_construct_uint32(OSSL_KDF_PARAM_ARGON2_LANES, &parallel_cost);
+    /* Set memory requirement */
+    *p++ = OSSL_PARAM_construct_uint32(OSSL_KDF_PARAM_ARGON2_MEMCOST, &memory_cost);
+    *p = OSSL_PARAM_construct_end();
+
+    if (EVP_KDF_CTX_set_params(kctx, params) != 1) {
+        fprintf(stderr, "EVP_KDF_CTX_set_params() failed\n");
+        goto end;
+    }
+
+    /* Derive the key */
+    if (EVP_KDF_derive(kctx, out, sizeof(out), NULL) != 1) {
```

Comment:
Suggestion: consider passing the `params` array to the derive call instead of separately via `EVP_KDF_CTX_set_params` above?